### PR TITLE
Proof-of-concept of likelihood "compression"

### PR DESCRIPTION
**Pull request by maxbiostat**
@csgillespie  see if you're happy with how I wrote the changes. I tried populating `xv` and `xf` with values from `internal` but it didn't quite work. Maybe that'd be an extra speed up (negligible compared to the actual compression). 
I went through `displ` but didn't see as many opportunities for compression (I think the likelihood already relies on a summary statistic).

I'm happy to write some more stuff if you can point the way. For really big data sets I reckon these "compression" tricks will bring nice speedups.

### CODE REVIEW

File: R/def_dislnorm.R [hunk 1]
```diff
@@ -208,16 +211,18 @@ setMethod("dist_rand",
 dislnorm$methods(
   mle = function(set = TRUE, initialise=NULL) {
     n = internal[["n"]]
```
Comment (csgillespie) [inline R/def_dislnorm.R]: We don't use `n` - so remove this line

File: R/def_dislnorm.R [hunk 2]
```diff
@@ -208,16 +211,18 @@ setMethod("dist_rand",
...
-    x = dat
-    x = x[x > (xmin-0.5)]
+    x = dat[dat > (xmin-0.5)]
+    tab = table(x)
+    xf = as.vector(tab)
+    xv = as.numeric(names(tab))
     x.log = log(x)
```
Comment (csgillespie) [inline R/def_dislnorm.R]: We should only take logs of x if we need to. So something like
```
    if(is.null(initialise)) {
      x = dat[dat > (xmin-0.5)]
      x.log = log(x)
      theta_0 = c(mean(x.log), sd(x.log))
    } else { 
      theta_0 = initialise
    }
```

File: R/def_dislnorm.R [hunk 3]
```diff
@@ -147,27 +147,30 @@ setMethod("dist_ll",
           definition = function(m) {
             xmin = m$getXmin()
             d = m$getDat()
-            dis_lnorm_tail_ll(d[d >= xmin], m$getPars(), xmin)
+            tab = table(d[d >= xmin])
```
Comment (csgillespie) [inline R/def_dislnorm.R]: See comment below on using internal for extra speed.

**Comment by csgillespie** [comment]
Hope these suggestions make sense? Your tweaks give a very nice speed improvement.

**Comment by maxbiostat** [comment]
Thanks! I'll work on the changes soon. I noticed it wasn't passing the tests and figured out why and wanted to wait until you reviewed the PR. For some reason when I tried to pull stuff from `internal`  I didn't get things to work. Anyway, will work on the fixes.

Are there any other distributions that could benefit from "data compression"?

**Comment by csgillespie** [comment]
- Data sets: the tests have gradually grown over the years. Bootstrapping occassionally generates silly data sets, e.g. c(1, 1)
- Data compression: I think any discrete distribution?
- Other speed-up: On my list of things to do was to allow users to enter discrete data as xv & xv instead of just x. That would allow the package to handle much large data sets. However, I think this would affect a large number of functions. Essentially, anywhere I use the data, I would need to check if it's discrete.

File: R/def_dislnorm.R (hunk 1 above)
Comment (maxbiostat) [reply R/def_dislnorm.R]: just to be clear, that line was there before me.

Comment (csgillespie) [reply R/def_dislnorm.R]: Not sure why I had it in. Just delete. Thanks

**Comment by maxbiostat** [comment]
@csgillespie, working on the PR for `dislnorm`. Will get to the others we have this one pitch-perfect. A doubt, though. In your opinion, which would be preferrable:
```R
   if(length(xv) == 1){
      joint_prob = sum(xf * log(apply(pars, 1, p)))
   }else{
     joint_prob = colSums(xf * log(apply(pars, 1, p)))
   }
```
or 
```R
 joint_prob = colSums( as.matrix( rbind(xf * log(apply(pars, 1, p)), 0) ) )
```
?
I went with the latter, but let me know.

**Comment by csgillespie** [comment]
@maxbiostat I think the `if()` statement is clearer. The one liner makes my head hurt trying to understand it ;)

**Comment by maxbiostat** [comment]
@csgillespie . Done. Would you like to merge this now or should I just look to implement compression for the other discrete distributions?

**Pull request merged**